### PR TITLE
Complete auth on server

### PR DESCRIPTION
**Pull request by RobertWSaunders**
## Proposed Change

This implements our auth endpoints:

- `oauth/session` 
- `oauth/signup`
- `oauth/refresh`
- `oauth/createResetHash`
- `oauth/updatePasswordForReset`

This change also includes schema changes to support those endpoints as well as various error related things to improve how we handle errors on the server. Changes to follow this one:

- Write tests for each endpoint and the controllers associated. (low priority)
- Improve how scopes are checked in resolvers. (medium priority)
- Handle Sequelize errors much better by mapping them to our own errors. (medium priority)
- Update how we handle dates on the server using moment. (medium priority)
- Make the frontend utilize these endpoints and make Apollo auto refresh tokens. (high priority)
- Refactor how we do emails and send emails for password reset (high priority)

Please see the comments in the PR for things I want to talk about or highlight for reference.

### How did you do this?

Using the Sequelize API and our new data model with thanks from @joeytepp! 😎 

### Why are you choosing this approach?

Add authentication and authorization into our app! 😏 

### Any open questions you want to ask code reviewers?

See the comments!

### List of changes:

  - Turn on Sequelize logging only for development
  - Enforce `clientId` on OAuthRefreshToken
  - Change `expires` to `expiryDate` in OAuthRefreshToken for consistency
  - Change relationship between OAuthUser and OAuthRefreshToken to a one-to-many one
  - Add `resetPasswordHashExpiryDate` in User
  - Add a bunch of custom errors and use them applicably
  - Add a missed association with OAuthUser to User
  - Authenticate every GraphQL request and set context
  - Removed some not used validation methods

## Pull Request Checklist:

  - [x] My submission passes all tests.
  - [x] I have lint my code locally prior to submission.
  - [x] I have written new tests for my core changes (where applicable).
  - [x] I have referenced all useful information (issues, tasks, etc) in the pull request.

### CODE REVIEW

File: packages/server/index.js [hunk 1]
```diff
@@ -115,10 +109,18 @@ db.sequelize
   .then(() => {
     logger.info("Database has synchronized successfully!");
 
-    // Start listening!
-    app.listen(port, () =>
-      logger.info(`QHacks Dashboard is running on port ${port}!`)
-    );
+    // NOTE: Temporary
+    return db.OAuthClient.create({
```
Comment (RobertWSaunders) [inline packages/server/index.js]: Should we keep this for now or not?

File: packages/server/oauth/controllers/oauth.controller.js [hunk 2]
```diff
@@ -1,68 +1,315 @@
...
+      });
+
+      return Promise.resolve({
+        scopes: newUser.scopes,
+        accessToken: newUser.accessToken,
+        refreshToken: newUser.refreshToken,
+        expiresIn: ACCESS_TOKEN_EXPIRE_SECONDS
+      });
+    } catch (err) {
+      return Promise.reject(new ValidationError("Database validation failed!"));
```
Comment (RobertWSaunders) [inline packages/server/oauth/controllers/oauth.controller.js]: Need to improve these errors!

File: packages/server/oauth/controllers/oauth.controller.js [hunk 3]
```diff
@@ -1,68 +1,315 @@
...
+          new ValidationError("Database validation failed!")
+        );
+      }
+      return Promise.reject(err);
+    }
+  }
+
+  async function refresh(hostname, refreshToken) {
+    return jwt.verify(refreshToken, AUTH_SECRET, async (err, decoded) => {
+      if (err) return Promise.reject(new OAuthApiError());
```
Comment (RobertWSaunders) [inline packages/server/oauth/controllers/oauth.controller.js]: Need to fix these lines, `OAuthApiError()` no longer exists! Note to self! 😃

File: packages/server/db/models/oauth-client.js [hunk 4]
```diff
@@ -51,7 +51,7 @@ module.exports = (sequelize, DataTypes) => {
     OAuthClient.hasMany(OAuthRefreshToken, {
       foreignKey: {
         name: "clientId",
-        allowNull: true
+        allowNull: false
```
Comment (joeytepp) [inline packages/server/db/models/oauth-client.js]: Nice 👍

File: packages/server/oauth/controllers/oauth.controller.js (hunk 3 above)
Comment (joeytepp) [reply packages/server/oauth/controllers/oauth.controller.js]: Sounds good!

File: packages/server/index.js (hunk 1 above)
Comment (RobertWSaunders) [reply packages/server/index.js]: I say we get rid of it for now!

**Pull request merged**